packages/core/src/machines/index.ts

```diff
@@ -385,23 +302,106 @@ export const createAuthMachine = ({
             }
           }
         },
-        email: {
-          initial: 'unknown',
+        registration: {
+          initial: 'incomplete',
           on: {
-            SIGNED_IN: [
-              {
-                cond: 'needsVerification',
-                target: '.awaitingVerification'
-              },
-              '.valid'
-            ],
-            SIGNOUT: '.unknown',
-            AWAIT_EMAIL_VERIFICATION: '.awaitingVerification'
+            SIGNED_IN: [{ cond: 'isAnonymous', target: '.incomplete' }, '.complete']
           },
           states: {
-            unknown: {},
-            awaitingVerification: {},
-            valid: {}
+            incomplete: {
+              on: {
+                SIGNUP_EMAIL_PASSWORD: 'emailPassword',
+                PASSWORDLESS_EMAIL: 'passwordlessEmail',
+                PASSWORDLESS_SMS: 'passwordlessSms',
+                PASSWORDLESS_SMS_OTP: 'passwordlessSmsOtp'
+              },
+              initial: 'noErrors',
+              states: {
+                noErrors: {},
+                needsEmailVerification: {},
+                needsOtp: {},
+                failed: {}
+              }
+            },
+            emailPassword: {
+              entry: ['resetErrors'],
+              invoke: {
+                src: 'signUpEmailPassword',
+                id: 'signUpEmailPassword',
+                onDone: [
+                  {
+                    cond: 'hasSession',
+                    actions: ['saveSession', 'reportTokenChanged'],
+                    target: '#nhost.authentication.signedIn'
+                  },
+                  {
+                    actions: 'clearContext',
+                    target: ['#nhost.authentication.signedOut', 'incomplete.needsEmailVerification']
+                  }
+                ],
+                onError: [
+                  {
+                    cond: 'unverified',
+                    target: 'incomplete.needsEmailVerification'
+                  },
+                  {
+                    actions: 'saveRegistrationError',
+                    target: 'incomplete.failed'
```

Comment:
I don't think we should treat failed registration as incomplete. A user would restart the whole registration workflow if there was an error during the process, wouldn't they?

`registration.failed` or `registration.error` makes more sense to me here

Reply:
If refactoring the logic here would require too many changes, then we can come back to this at a later time.

Reply:
The logic is: a failed registration is incomplete.
The main reason why it is structured that way is to share transitions with other incomplete sub-states: `SIGNUP_EMAIL_PASSWORD`, `PASSWORDLESS_EMAIL` etc.
These transitions are on the `incomplete` state so we avoid running them when a registration step is already ongoing (for instance `passwordlessEmail`), or when registration is complete.

We could copy-paste every transition in a new `failed` state and refactor, but I'd say we keep it that way and we can change this on a separate PR

Reply:
To me an incomplete state would be a state where we already have some progress, but has not been finished for some reason (such as the email verification). I agree that we should not necessarily change it in this PR.

It'd also be great to see the opinion of @elitan / @nunopato regarding this.

Reply:
I don't know what "incomplete" means in this context.

I don' know what "share transitions with other incomplete sub-states" means.

Generally, I'm so lost in this xstate / machine code to the point that I can not contribute with anything meaningful. Sorry. 

I feel this is over-engineered. But that's a different topic.

Reply:
I'd suggest we should look into this issue later. I understand your concern about `xstate`, especially regarding the complexity it brings seemingly, although I feel that it helps a lot with transitions between different states of the SDK. We should definitely look for ways to make the code more compact.